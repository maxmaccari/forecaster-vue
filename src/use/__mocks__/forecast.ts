export const useForecast = jest.fn()