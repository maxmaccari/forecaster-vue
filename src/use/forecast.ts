declare interface WindResponse {
  speed: number
  deg: number
}

declare interface WeatherResponse {
  coord: { lon: number; lat: number }
  weather: {
    id: number
    main: string
    description: string
    icon: string
  }[]
  base: string
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    humidity: number
  }
  visibility: number
  wind: WindResponse
  clouds: {
    all: number
  }
  dt: number
  sys: {
    type: number
    id: number
    country: string
    sunrise: number
    sunset: number
  }
  timezone: number
  id: number
  name: string
  cod: number
}

declare interface ListResponse {
  dt: number
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    sea_level: number
    grnd_level: number
    humidity: number
    temp_kf: number
  }
  weather: [
    {
      id: number
      main: string
      description: string
      icon: string
    }
  ]
  clouds: { all: number }
  wind: WindResponse
  visibility: number
  pop: number
  sys: { pod: string }
  dt_txt: string
}

declare interface ForecastResponse {
  cod: string
  message: number
  cnt: number
  list: ListResponse[]
  city: {
    id: number
    name: string
    coord: { lat: number; lon: number }
    country: string
    population: number
    timezone: number
    sunrise: number
    sunset: number
  }
}

export class Weather {
  description: string | null = null
  icon: string | null = null
  temperature = 0
  minTemperature = 0
  maxTemperature = 0
  feelsLike = 0
  humidity = 0
  pressure = 0

  constructor(weather: WeatherResponse | ListResponse | null = null) {
    if (!weather) return

    this.description = weather.weather[0]
      ? weather.weather[0].description
      : null
    this.icon = weather.weather[0] ? weather.weather[0].icon : null
    this.temperature = weather.main.temp
    this.minTemperature = weather.main.temp_min
    this.maxTemperature = weather.main.temp_max
    this.feelsLike = weather.main.feels_like
    this.humidity = weather.main.humidity
    this.pressure = weather.main.pressure
  }
}

export class Wind {
  direction = 0
  speed = 0

  constructor(windResponse: WindResponse | null = null) {
    this.direction = windResponse?.deg || 0
    this.speed = windResponse?.speed || 0
  }
}

export class ForecastDetails {
  clouds = 0
  timestamp = 0
  weather: Weather = new Weather()
  wind: Wind = new Wind()

  constructor(details: ListResponse) {
    this.clouds = details.clouds.all
    this.timestamp = details.dt
    this.weather = new Weather(details)
  }
}

export class Forecast {
  name = ''
  country = ''
  weather: Weather = new Weather()
  wind: Wind = new Wind()
  details: ForecastDetails[] = []

  constructor(weather: WeatherResponse, forecast: ForecastResponse) {
    this.name = weather.name
    this.country = forecast.city.country
    this.weather = new Weather(weather)
    this.wind = new Wind(weather.wind)
    this.details = forecast.list.map(details => {
      return new ForecastDetails(details)
    })
  }
}

class CacheHash {
  [Identifier: string]: Forecast
}

let cache: CacheHash = new CacheHash()

const API_URL = process.env.VUE_APP_API_URL
const API_KEY = process.env.VUE_APP_API_KEY

export const fetchWeather = (location: string): Promise<WeatherResponse> => {
  const url = `${API_URL}/weather?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as WeatherResponse
  })
}

export const fetchForecast = (location: string): Promise<ForecastResponse> => {
  const url = `${API_URL}/forecast?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as ForecastResponse
  })
}

export const clearCache = () => {
  cache = new CacheHash()
}

export const useForecast = async (location: string): Promise<Forecast> => {
  if (cache[location]) return cache[location]

  const [weatherJson, forecastJson] = await Promise.all([
    fetchWeather(location),
    fetchForecast(location),
  ])

  const forecast = new Forecast(weatherJson, forecastJson)
  cache[location] = forecast

  return forecast
}
