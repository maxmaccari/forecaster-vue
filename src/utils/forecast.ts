import { Forecast } from '@/use/forecast'

export class NextHoursSummary {
  constructor(
    public temperature: number,
    public time: string,
    public clouds: number,
    public icon: string | null
  ) {}
}

export class NextWeekEntry {
  constructor(
    public day: string,
    public clouds: number,
    public min: number,
    public temperature: number,
    public icon: string | null,
    public description: string | null
  ) {}
}

export const dateFormat = new Intl.DateTimeFormat('en-US', {
  day: 'numeric',
  weekday: 'long',
})

export const timeFormat = new Intl.DateTimeFormat('en-US', {
  hour: 'numeric',
  minute: 'numeric',
})

export const getNextHours = (forecast: Forecast): NextHoursSummary[] => {
  return [forecast.details[0], forecast.details[2], forecast.details[4]].map(
    detail => new NextHoursSummary(
      Math.round(detail.weather.temperature),
      timeFormat.format(new Date(detail.timestamp * 1000)),
      detail.clouds,
      detail.weather.icon,
    )
  )
}

export const getNextWeek = (forecast: Forecast): NextWeekEntry[] => {
  return forecast.details.reduce((week: NextWeekEntry[], detail) => {
    const date = new Date(detail.timestamp * 1000)
    const day = dateFormat.format(date)
    const forecast = week.find(forecast => forecast.day === day)

    if (!forecast) {
      week.push(new NextWeekEntry(day, 
        detail.clouds,
        Math.round(detail.weather.minTemperature),
        Math.round(detail.weather.temperature),
        detail.weather.icon,
        detail.weather.description,
      ))
    } else {
      if (detail.weather.minTemperature < forecast.min) {
        forecast.min = Math.round(detail.weather.minTemperature)
      }

      if (detail.weather.temperature > forecast.temperature) {
        forecast.temperature = Math.round(detail.weather.temperature)
      }

      if (date.getHours() > 10 && date.getHours() < 14) {
        forecast.icon = detail.weather.icon
      }
    }

    return week
  }, [])
}
