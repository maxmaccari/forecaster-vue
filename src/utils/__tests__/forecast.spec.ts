import { describe, it, expect } from 'vitest'

import { buildForecast } from '@unit/helpers/forecast'
import { getDetailsWeather, getNextHours, getNextWeek, dateFormat, getDetailsNextHours } from '../forecast'

const forecast = buildForecast()
const randomDateFromForecast = () => {
  const days = forecast.details.map(detail => 
    new Date(detail.timestamp * 1000).toISOString().slice(0, 10)
  );
  const dateDays = [...new Set(days)]
    .map(date => new Date(`${date}T00:00:00`))

  return dateDays[Math.floor(Math.random() * dateDays.length)];
}

describe('getNextHours', () => {
  it('should return the next hours data based on the forecast data', function() {
    expect(getNextHours(forecast)).toMatchSnapshot()
  })
})

describe('getNextWeek', () => {
  it('should return the next week data based on the forecast data', function() {
    expect(getNextWeek(forecast)).toMatchSnapshot()
  })
})

describe('getDetailsWeather', () => {
  it('should return the weather from the day based on the forecast data', function() {
    const date = randomDateFromForecast();
    const weather = getDetailsWeather(forecast, date);

    expect(weather.isoDate).toBe(date.toISOString().slice(0, 10))
  })
})

describe('getDetailsNextHours', () => {
  it('should return the forecasts for the given day', function() {
    const date = randomDateFromForecast();
    const nextHoursDetails = getDetailsNextHours(forecast, date);
    const nextHoursDetailsDays = nextHoursDetails
      .map(detail => dateFormat.format(new Date(detail.timestamp * 1000)));

    for (const day of nextHoursDetailsDays) {
      expect(day).toBe(dateFormat.format(date))
    }
  })
})