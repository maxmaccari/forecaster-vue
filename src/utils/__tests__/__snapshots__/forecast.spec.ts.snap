// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getNextHours > should return the next hours data based on the forecast data 1`] = `
[
  NextHoursSummary {
    "clouds": 100,
    "icon": "021-cloud",
    "temperature": 11,
    "time": "6:00 PM",
  },
  NextHoursSummary {
    "clouds": 52,
    "icon": "021-cloud",
    "temperature": 10,
    "time": "12:00 AM",
  },
  NextHoursSummary {
    "clouds": 0,
    "icon": "021-sun",
    "temperature": 14,
    "time": "6:00 AM",
  },
]
`;

exports[`getNextWeek > should return the next week data based on the forecast data 1`] = `
[
  WeatherSummary {
    "clouds": 100,
    "date": 2020-09-27T21:00:00.000Z,
    "description": "overcast clouds",
    "feelsLike": 9,
    "humidity": 79,
    "icon": "021-cloud",
    "max": 12,
    "min": 11,
    "pressure": 1007,
    "temperature": 11,
    "wind": Wind {
      "direction": 204,
      "speed": 2.85,
    },
  },
  WeatherSummary {
    "clouds": 52,
    "date": 2020-09-28T03:00:00.000Z,
    "description": "broken clouds",
    "feelsLike": 8,
    "humidity": 76,
    "icon": "021-cloud",
    "max": 10,
    "min": 10,
    "pressure": 1009,
    "temperature": 18,
    "wind": Wind {
      "direction": 192,
      "speed": 1.9,
    },
  },
  WeatherSummary {
    "clouds": 99,
    "date": 2020-09-29T03:00:00.000Z,
    "description": "overcast clouds",
    "feelsLike": 11,
    "humidity": 71,
    "icon": "021-cloud",
    "max": 13,
    "min": 12,
    "pressure": 1014,
    "temperature": 19,
    "wind": Wind {
      "direction": 316,
      "speed": 1.02,
    },
  },
  WeatherSummary {
    "clouds": 11,
    "date": 2020-09-30T03:00:00.000Z,
    "description": "few clouds",
    "feelsLike": 12,
    "humidity": 80,
    "icon": "021-sun",
    "max": 12,
    "min": 12,
    "pressure": 1015,
    "temperature": 20,
    "wind": Wind {
      "direction": 248,
      "speed": 0.68,
    },
  },
  WeatherSummary {
    "clouds": 89,
    "date": 2020-10-01T03:00:00.000Z,
    "description": "overcast clouds",
    "feelsLike": 11,
    "humidity": 73,
    "icon": "021-cloudy-1",
    "max": 13,
    "min": 13,
    "pressure": 1010,
    "temperature": 20,
    "wind": Wind {
      "direction": 141,
      "speed": 1.11,
    },
  },
  WeatherSummary {
    "clouds": 100,
    "date": 2020-10-02T03:00:00.000Z,
    "description": "overcast clouds",
    "feelsLike": 11,
    "humidity": 73,
    "icon": "021-cloud",
    "max": 14,
    "min": 13,
    "pressure": 1001,
    "temperature": 17,
    "wind": Wind {
      "direction": 223,
      "speed": 4.51,
    },
  },
]
`;
